### DISCUSSION TITLE
Relay trace from offline service to collector

### DISCUSSION THREAD

**Discussion by NullHypothesis** [open-telemetry/opentelemetry-go#5051] [General]
I have a Go service that spawns a Docker container, which runs a Python service. Both services are instrumented via OpenTelemetry. The Python service however is sandboxed and does not have network connectivity. It receives the Go service's trace context via environment variable and can communicate back via text files written to disk. 

```mermaid
sequenceDiagram
    participant Collector
    participant Go svc
    participant Python svc
    Go svc->>Go svc: Start trace
    Go svc->>Python svc: Start container (TRACE_CTX=...)
    Python svc->>Python svc: Continue trace
    Python svc->>Go svc: Trace (as text file)
    Go svc->>Collector: Export traces
```

So far, the Python service uses the [`ConsoleSpanExporter`](https://opentelemetry-python.readthedocs.io/en/latest/sdk/trace.export.html?highlight=consolespanexporter#opentelemetry.sdk.trace.export.ConsoleSpanExporter) to write its trace to a file on disk, which makes it back to the Go service. I now wonder what's the easiest way to have the Go service take this text file and send it to the collector. Any suggestions?

**Comment by dmathieu** [comment]
The ConsoleSpanExporter is there for debugging purposes, not for production use. There is no formalized protocol and its output could have breaking changes.
So if you really wish to read from it, you will have to write your own parser.

OTLP is actually the protocol built for exporting traces. So I would consider either of the following:

* Expose a collector to your Python service, so it can send traces using OTLP.
* Use [opentelemetry-proto-go](https://github.com/open-telemetry/opentelemetry-proto-go) to build a new endpoint in your Go service, which would receive OTLP data and forward it to the collector.

For both solutions, the view of the Python service remains the same: it sends OTLP data. Whether that's through a local collector (which can be dedicated to that), or with a custom built solution in your service is up to you.
